File: chap18/Addition.java
// Ces programmes sont sous licence CeCILL-B V1.
class Addition {

  public static void main (String [] args) {
    boolean [] n, p, r;
    int i;
    boolean a,b,c;
    n = new boolean [10];
    p = new boolean [10];
    r = new boolean [11];
    n[0] = true;
    n[1] = true;
    n[2] = true;
    n[3] = false;
    n[4] = false;
    n[5] = false;
    n[6] = false;
    n[7] = false;
    n[8] = false;
    n[9] = false;
    p[0] = false;
    p[1] = true;
    p[2] = false;
    p[3] = false;
    p[4] = false;
    p[5] = false;
    p[6] = false;
    p[7] = false;
    p[8] = false;
    p[9] = false;
    c = false;
    for (i = 0; i <= 9; i = i + 1) {
      a = n[i];
      b = p[i];
      r[i] = (a && !b && !c) || (!a && b && !c) || (!a && !b && c) 
          || (a && b && c);
      c = (a && b) || (b && c) || (a && c);}
    r[10] = c;
    System.out.print(" "); 
    for (i = 0; i <= 9; i = i + 1) {
      if (n[9-i]) {
        System.out.print("1");} 
      else {
        System.out.print("0");}}
    System.out.println();
    System.out.print(" "); 
    for (i = 0; i <= 9; i = i + 1) {
      if (p[9-i]) {
        System.out.print("1");} 
      else {
        System.out.print("0");}}
    System.out.println();
    for (i = 0; i <= 10; i = i + 1) {
      if (r[10-i]) {
        System.out.print("1");} 
      else {
        System.out.print("0");}}
    System.out.println();}}

File: chap18/Addition.ml
(* Ces programmes sont sous licence CeCILL-B V1. *)

let _ =
    let n = Array.create 10 false in
    let p = Array.create 10 false in
    let r = Array.create 11 true in
    n.(0) <- true;
    n.(1) <- true;
    n.(2) <- true;
    p.(1) <- true;

    let c = ref false in
    for i = 0 to 9 do
      let a = n.(i) in
      let b = p.(i) in
      r.(i) <- (a && (not b) && (not !c)) || ((not a) && b && (not !c))
      || ((not a) && (not b) && !c) || (a && b && !c);
      c := (a && b) || (b && !c) || (a && !c);
    done;
    r.(10) <- !c;
    print_string " ";
    for i = 0 to 9 do
      if n.(9-i) then
        print_string "1"
      else
        print_string "0"
    done;
    print_newline ();
    print_string " ";
    for i = 0 to 9 do
      if p.(9-i) then
        print_string "1"
      else
        print_string "0"
    done;
    print_newline ();
    for i = 0 to 10 do
      if r.(10-i) then
        print_string "1"
      else
        print_string "0";
    done;
    print_newline ()

